### PR TITLE
Driver networking support

### PR DESCRIPTION
**Pull request by nickethier**
Adds support for passing network isolation config into drivers and
implements support in the rawexec driver as a proof of concept

### CODE REVIEW

File: drivers/shared/executor/executor.go [hunk 1]
```diff
@@ -303,8 +306,30 @@ func (e *UniversalExecutor) Launch(command *ExecCommand) (*ProcessState, error)
 	e.childCmd.Env = e.commandCfg.Env
 
 	// Start the process
-	if err := e.childCmd.Start(); err != nil {
-		return nil, fmt.Errorf("failed to start command path=%q --- args=%q: %v", path, e.childCmd.Args, err)
+	if command.NetworkIsolation != nil && command.NetworkIsolation.Path != "" {
+		// Lock to the thread we're changing the network namespace of
+		runtime.LockOSThread()
```
Comment (notnoop) [inline drivers/shared/executor/executor.go]: Is this necessary?  Seeing `ns` code, I see that the `Do` function is called in a different goroutine and that goroutine is already protected with `runtime.LockOSThread()` in https://github.com/containernetworking/plugins/blob/fe60fcddb897079746ec1523fd1837ab05b1e689/pkg/ns/ns_linux.go#L197-L201 ; and I don't see runtime call in the CNI example in https://github.com/hwchiu/CNI_Tutorial_2018/blob/ea43c401127955a1e8974cfd03829509c17f214a/tutorial/step3/example.go .

If needed, let's comment the need and also point to why we shouldn't be calling `runtime.UnlockOSThread()`.

I

File: drivers/shared/executor/executor.go [hunk 2]
```diff
@@ -303,8 +306,30 @@ func (e *UniversalExecutor) Launch(command *ExecCommand) (*ProcessState, error)
...
+		err = netns.Do(func(ns.NetNS) error {
+
+			if err := e.childCmd.Start(); err != nil {
+				return fmt.Errorf("failed to start command path=%q --- args=%q: %v", path, e.childCmd.Args, err)
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
```
Comment (notnoop) [inline drivers/shared/executor/executor.go]: nit pick, you probably can simplify this a bit with something like:

```go
		err = netns.Do(func(ns.NetNS) error { return e.childCmd.Start() }
		if err != nil {
			return nil,  fmt.Errorf("failed to start command path=%q --- args=%q: %v", path, e.childCmd.Args, err)
		}
```

File: drivers/shared/executor/executor.go [hunk 3]
```diff
@@ -303,8 +306,30 @@ func (e *UniversalExecutor) Launch(command *ExecCommand) (*ProcessState, error)
...
 
 	// Start the process
-	if err := e.childCmd.Start(); err != nil {
-		return nil, fmt.Errorf("failed to start command path=%q --- args=%q: %v", path, e.childCmd.Args, err)
+	if command.NetworkIsolation != nil && command.NetworkIsolation.Path != "" {
+		// Lock to the thread we're changing the network namespace of
+		runtime.LockOSThread()
+		netns, err := ns.GetNS(command.NetworkIsolation.Path)
+		if err != nil {
+			return nil, err
```
Comment (notnoop) [inline drivers/shared/executor/executor.go]: nitpick: might be nice to wrap this error to get better logs, e.g.
```suggestion
			return nil, fmt.Errof("failed to get ns %s: %v", command.NetworkIsolation.Path, err)
```

File: drivers/shared/executor/executor_linux.go [hunk 4]
```diff
@@ -195,9 +197,30 @@ func (l *LibcontainerExecutor) Launch(command *ExecCommand) (*ProcessState, erro
...
+		runtime.LockOSThread()
+		netns, err := ns.GetNS(command.NetworkIsolation.Path)
+		if err != nil {
+			return nil, err
+		}
+
+		// Start the container in the network namespace
+		err = netns.Do(func(ns.NetNS) error {
+			if err := container.Run(process); err != nil {
+				container.Destroy()
```
Comment (notnoop) [inline drivers/shared/executor/executor_linux.go]: Q: Does the destroy also need to happen in the netns.Do context?  If so, we need to wrap other `container.Destroy` calls?  If not, we probably can reduce if statements for error handling like above.

File: drivers/shared/executor/executor.go (hunk 1 above)
Comment (nickethier) [reply drivers/shared/executor/executor.go]: You are correct, I think I talked about this with someone else too. Calling `runtime.LockOSThread()` is sort of a muscle memory I've developed from working with network namespaces before this `ns.Do` block existed.

**Pull request merged**